### PR TITLE
Fix SPI datum allocation memory leaks

### PR DESCRIPTION
**Pull request by eeeebbbbrrrr**
The intent of SPI has been that it use whatever `CurrentMemoryContext` is prior to calling `Spi::connect()` to allocate Datums.  This allows the `Spi::connect()` closure to return type instances that might be backed by Postgres-allocated memory (`&str` and `&[u8]`, specifically).

While operating in Spi-land, we're inside a short-lived "SPI Proc" memory context -- we enter it upon `SPI_connect()` and it's free'd at `SPI_finish()`.  Unfortunately, an incorrect assumption was made that this memory context's parent is `CurrentMemoryContext` at the time `SPI_connect()` is called.  This is not the case.

Postgres attaches this "SPI Proc" context to `TopTransactionContext` instead.  As such, pgrx' use of `PgMemoryContexts::CurrentMemoryContext.parent()` to find the context for datum allocation is the wrong one.

This PR adjusts the Spi wrapper internals such that the `SpiClient` holds onto a `PgMemoryContext` and that it is set to `CurrentMemoryContext` **before** we call `SPI_connect().`

This resolves the first complaint from https://github.com/pgcentralfoundation/pgrx/issues/1209#issuecomment-1635614847 that Spi is leaky.

### CODE REVIEW

File: pgrx/src/spi.rs [hunk 1]
```diff
@@ -232,8 +232,26 @@ impl Spi {
...
+}
+
+impl DatumMemoryContext {
+    #[inline]
+    fn to_inner(&self) -> PgMemoryContexts {
+        self.clone().0
+    }
+}
+
 // TODO: should `'conn` be invariant?
```
Comment (workingjubilee) [inline pgrx/src/spi.rs:251]: an excellent question, January Eric.

**Comment by david-monroe** [comment]
If this allocates into the current function context, then it still disables the hack mentioned in https://docs.rs/pgrx/latest/pgrx/spi/struct.SpiCursor.html#important-notes-about-memory-usage.

File: pgrx/src/spi.rs (hunk 1 above)
Comment (workingjubilee) [reply pgrx/src/spi.rs:251]: The answer is **extremely** yes.

**Comment by eeeebbbbrrrr** [comment]
I don't believe this is going to make it, and that's fine!

**Pull request closed without merging**